### PR TITLE
Fix Smac Hybrid search state after costmap resolution changes

### LINKED ISSUE #4981: SMAC hybrid not following the minimum turning radius set
## Bug report
The turning radius is not as expected, I think it is a bug related to the minimum_turning_radius in the SMAC hybrid planner for ROS:Iron.
I looked into the problem, it seems not related to my current code, although I might have missed something that will affect the turning radius.
Thank you in advance for any insight into the problem or the bug.

**Required Info:**
- Operating System:
Docker, Ubuntu 22 
- ROS2 Version:
ROS2 Iron
- Version or commit hash:
sudo apgt install for docker >>   ros-${ROS_DISTRO}-navigation2 

#### Steps to reproduce issue
Load a simulation, e.g. turtlebot, or preferably an Ackermann robot sim. 
Set the planner smac hybrid and inspect the turning radius corresponding to the actual distances on a map.
#### Expected behavior
Turning radius modifications make the planner take bigger turners, or only as small as the minimum turning radius.
The turning radius when set to a radius in meters corresponds to that radius on the map.

#### Actual behavior
Parameter seems to affect the turning, but not by the set radius.

#### Reproduction instructions
Change the minimum_turning_radius parameter from the planner: SMAC hybrid.

#### Additional information
Ackerman cart, fixing the in proper paths through the controller, is not fixing the issue as of yet.

### PR DESCRIPTION
**Pull request by mini-1235**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | Fixes https://github.com/ros-navigation/navigation2/issues/4981 |
| Primary OS tested on | MacOS |
| Robotic platform tested on | Planner Playground |
| Does this PR contain AI generated software? | Yes, Codex 5.6 Sol |
| Was this PR description generated by AI software? | No |

---

## Description of how this change was tested

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists
- [ ] Should this be backported to current distributions? If so, tag with `backport-*`.

### CODE REVIEW

**Comment by mini-1235** [comment]
@SteveMacenski If you still remember our conversation from a few months ago, I narrowed the issue down to costmap resolution, but I never figured out the "root cause" in the code.

I think I understand it now after reading through the costmap code today (mainly looking for some inspiration for my own work). To summarize, I believe the issue mostly happens when the static map resolution differs from the `resolution` ROS parameter:

```text
Costmap configures at 0.10 (ROS parameter)
Smac configures → reads 0.10 and caches values
StaticLayer receives map → costmap becomes 0.05
Planning → Smac continues using stale cached values
```

The tl;dr is that I think there is an inconsistency in the resolution used throughout Smac Planner. I came up with a fix and verified it in the planner playground.

6b7c23b4b7dcb9ec6fbc7e39bae1eca2

The green path is before this PR and the red path is after, with a minimum turning radius of 2.5. Based on the results from my planner playground script, I believe the red path is the correct one, which also looks more reasonable visually.

I probably still need some time to clean up linting, tests, etc., but wanted to get a quick review first to make sure you agree with the overall direction :)

**Comment by SteveMacenski** [comment]
Does the discussion in the ticket also validate that / the reporter's configurations or setups have the differing resolutions?

File: nav2_smac_planner/include/nav2_smac_planner/smac_planner_hybrid.hpp [hunk 1]
```diff
@@ -139,6 +148,8 @@ class SmacPlannerHybridT : public nav2_core::GlobalPlanner
   SearchInfo _search_info;
   double _max_planning_time;
   double _lookup_table_size;
+  double _search_resolution{0.0};
+  double _analytic_expansion_max_length_m;
```
Comment (SteveMacenski) [inline nav2_smac_planner/include/nav2_smac_planner/smac_planner_hybrid.hpp:152]: I'd prefer if this was not a class member; but I'm not sure if there's a reasonable way around that for how its used

File: nav2_smac_planner/include/nav2_smac_planner/smac_planner_hybrid_impl.hpp [hunk 2]
```diff
@@ -896,14 +883,14 @@ SmacPlannerHybridT<NodeT>::updateParametersCallback(
...
 
-    // Re-Initialize smoother
-    if (reinit_smoother) {
-      SmootherParams params;
-      params.get(node, _name);
-      _smoother = std::make_unique<Smoother>(params);
-      _smoother->initialize(_minimum_turning_radius_global_coords);
-    }
+  // Re-Initialize smoother
+  if (reinit_smoother) {
```
Comment (SteveMacenski) [inline nav2_smac_planner/include/nav2_smac_planner/smac_planner_hybrid_impl.hpp]: Why was this removed from the same scope as the others & `if (reinit_a_star || reinit_downsampler || reinit_collision_checker)`?

File: nav2_smac_planner/include/nav2_smac_planner/smac_planner_hybrid_impl.hpp [hunk 3]
```diff
@@ -369,9 +366,20 @@ nav_msgs::msg::Path SmacPlannerHybridT<NodeT>::createPlan(
...
+  // The resolution can change, even if the parameter is not updated (e.g. if the map is provided with a different resolution),
+  // Therefore, we need to check the costmap resolution and update the search resolution if it has changed.
+  if (_search_resolution != costmap->getResolution()) {
+    updateSearchResolution(costmap->getResolution());
+    _a_star = std::make_unique<AStarAlgorithm<NodeT>>(_motion_model, _search_info);
+    _a_star->initialize(
+      _allow_unknown, _max_iterations, _max_on_approach_iterations,
+      _terminal_checking_interval, _max_planning_time, _lookup_table_dim,
+      _angle_quantizations);
+  }
```
Comment (SteveMacenski) [inline nav2_smac_planner/include/nav2_smac_planner/smac_planner_hybrid_impl.hpp]: Can this just be moved before Downsample costmap so that we can keep the same `_collision_checker.setCostmap(costmap);` line?

**Review by SteveMacenski — COMMENTED** [review]
Seems sane!

**Comment by mini-1235** [comment]
> Does the discussion in the ticket also validate that / the reporter's configurations or setups have the differing resolutions?

I can't say with 100% certainty that the issues reported there were caused by the resolution mismatch fixed in this PR, since I don't have access to their exact setup and the issue report itself doesn't provide much detail. However, I think there are some supporting evidence:

- Based on your request, the original reporter said they could "repro the same issue with the planner playground" in https://github.com/ros-navigation/navigation2/issues/4981#issuecomment-2744041442. In the planner playground defaults, the costmap resolution is 0.10 while the map resolution is 0.05. My screenshot above was also taken using the planner playground :)

- The second person involved in the thread also appears to be using the planner playground: https://github.com/ros-navigation/navigation2/issues/4981#issuecomment-2830718643

- A third person explicitly mentioned using global_costmap.resolution: 0.25: https://github.com/ros-navigation/navigation2/issues/4981#issuecomment-4213321062

- I also tested multiple start/goal poses and different minimum turning radius using the tool linked here: https://github.com/ros-navigation/navigation2/issues/4981#issuecomment-3745710465. In most cases, I was no longer able to reproduce the path violation. I did notice a few other minor issues, which I plan to open follow-up PRs for discussion separately 

Overall, based on these observations and the result shown in my screenshot above, I believe this PR addresses the reported issue (?)

File: nav2_smac_planner/include/nav2_smac_planner/smac_planner_hybrid_impl.hpp (hunk 2 above)
Comment (mini-1235) [reply nav2_smac_planner/include/nav2_smac_planner/smac_planner_hybrid_impl.hpp]: This is a bit tricky .... 

I was trying to add
`std::unique_lock<nav2_costmap_2d::Costmap2D::mutex_t> lock(*(_costmap->getMutex()));`
to the dynamic parameter callback. I don't think `reinit_smoother` needs to read the costmap state, but `reinit_a_star`, `reinit_downsampler`, and `reinit_collision_checker` do (?), so I moved it out. I can also move it back if you would prefer to keep the logic together

File: nav2_smac_planner/include/nav2_smac_planner/smac_planner_hybrid_impl.hpp (hunk 3 above)
Comment (mini-1235) [reply nav2_smac_planner/include/nav2_smac_planner/smac_planner_hybrid_impl.hpp]: I think `_collision_checker.setCostmap(costmap)` should remain outside the block anyway. If we initially enable downsampling and later disable it, the collision checker must switch from the downsampled costmap back to the original costmap. Since `setCostmap()` only performs a pointer assignment, I don't this would hurt the performance

Same for Smac2D, I think we should also move it out

File: nav2_smac_planner/include/nav2_smac_planner/smac_planner_hybrid_impl.hpp (hunk 2 above)
Comment (SteveMacenski) [reply nav2_smac_planner/include/nav2_smac_planner/smac_planner_hybrid_impl.hpp]: I'd rather keep it altogether if there's not a technical reason it should be removed (just organizationally)

File: nav2_smac_planner/include/nav2_smac_planner/smac_planner_hybrid_impl.hpp (hunk 3 above)
Comment (SteveMacenski) [reply nav2_smac_planner/include/nav2_smac_planner/smac_planner_hybrid_impl.hpp]: Fair enough

**Comment by SteveMacenski** [comment]
Sounds good! Can you comment on the ticket regarding root cause & "@"ing the users to make sure this aligns with their setup to get a :+1: that this resolves the issue for all? 

Go ahead and finish this PR up and I'll merge :-)

File: nav2_smac_planner/include/nav2_smac_planner/smac_planner_hybrid_impl.hpp (hunk 2 above)
Comment (mini-1235) [reply nav2_smac_planner/include/nav2_smac_planner/smac_planner_hybrid_impl.hpp]: Ok, I don't think it makes any practical difference, moving it back

**Comment by mini-1235** [comment]
Ready to merge from my side!

**Review by SteveMacenski — APPROVED** [review]
